File: 2368-reachable-nodes-with-restrictions/2368-reachable-nodes-with-restrictions.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @param {number[]} restricted
 * @return {number}
 */
var reachableNodes = function(n, edges, restricted) {
    if (restricted.length === 0) return n
    if (restricted.includes(0)) return 0
    restricted = new Set(restricted)
    restricted.add(0)
    const graph = {}, q = [0]
    let visited = 0
    for (let i=0; i < n; i++) {
        graph[i] = []
    }
    for (const [u, v] of edges) {
        graph[u].push(v)
        graph[v].push(u)
    }
    while (q.length) {
        visited++
        const cur = q.shift()
        for (const next of graph[cur]) {
            if (!restricted.has(next)) {
                restricted.add(next)
                q.push(next)
            }
        }
    }
    return visited
};

File: 2368-reachable-nodes-with-restrictions/2368-reachable-nodes-with-restrictions.py
class Solution:
    def __init__(self):
        self.visited = 0
        self.graph = defaultdict(list)
        
    def dfs(self, root: int, restricted: set) -> None:
        self.visited += 1
        for next in self.graph[root]:
            if next not in restricted:
                restricted.add(next)
                self.dfs(next, restricted)
    
    def reachableNodes(self, n: int, edges: List[List[int]], restricted: List[int]) -> int:
        if len(restricted) == 0:
            return n
        if 0 in restricted:
            return 0
        for u, v in edges:
            self.graph[u].append(v)
            self.graph[v].append(u)
        restricted = set(restricted)
        restricted.add(0)
        self.dfs(0, restricted)
        return self.visited